File: Hard/Python/1326. Minimum Number of Taps to Open to Water a Garden.py
class Solution(object):
    def minTaps(self, n, ranges):
        this_range = [0] * (n + 1)

        for i in range(len(ranges)):
            left = max(0, i - ranges[i])
            right = min(n, i + ranges[i])
            this_range[left] = max(this_range[left], right)
        
        end, maxRange, taps, i = 0, 0, 0, 0

        while end < n:
            while i <= end:
                maxRange = max(maxRange, this_range[i])
                i += 1
            
            if maxRange <= end:
                return -1 # Can't water the garden
            end = maxRange
            taps += 1
        
        return taps

File: Hard/Java/1326. Minimum Number of Taps to Open to Water a Garden.java
class Solution {
    public int minTaps(int n, int[] ranges) {
        int[] range = new int[n + 1];

        for(int i = 0; i < ranges.length; i++){
            int left = Math.max(0, i - ranges[i]);
            int right = Math.min(n, i + ranges[i]);
            range[left] = Math.max(range[left], right);
        }

        int end = 0;
        int maxRange = 0;
        int taps = 0;
        int i = 0;

        while(end < n){
            while(i <= end){
                maxRange = Math.max(maxRange, range[i]);
                i++;
            }

            if(maxRange <= end)
                return -1; // Can't water the garden
            end = maxRange;
            taps++;
        }

        return taps;
    }
}
